README.md
# :satellite: xrit-rx - LRIT/HRIT Downlink Processor

[![GitHub release](https://img.shields.io/github/release/sam210723/xrit-rx.svg)](https://github.com/sam210723/xrit-rx/releases/latest)
[![Python versions](https://img.shields.io/badge/python-3.7%20%7C%203.8%20%7C%203.9%20%7C%203.10-blue)](https://www.python.org/)
[![Github all releases](https://img.shields.io/github/downloads/sam210723/xrit-rx/total.svg)](https://github.com/sam210723/xrit-rx/releases/latest)
[![GitHub license](https://img.shields.io/github/license/sam210723/xrit-rx.svg)](https://github.com/sam210723/xrit-rx/blob/master/LICENSE)

**xrit-rx** is a packet demultiplexer and file processor for receiving images from geostationary weather satellite [GEO-KOMPSAT-2A (GK-2A)](https://nmsc.kma.go.kr/enhome/html/base/cmm/selectPage.do?page=satellite.gk2a.intro). It is designed for use with [**SatDump**](https://github.com/altillimity/SatDump) by [Aang254](https://twitter.com/aang254), [**goesrecv**](https://github.com/sam210723/goestools) (originally by [Pieter Noordhuis](https://twitter.com/pnoordhuis)), or [**xritdecoder**](https://github.com/opensatelliteproject/xritdemod/releases/tag/1.0.3) by [Lucas Teske](https://twitter.com/lucasteske).

**xrit-rx** receives [Virtual Channel Data Units (VCDUs)](https://nmsc.kma.go.kr/resources/homepage/pdf/GK2A_LRIT_Mission_Specification_Document_v1.0.pdf#page=27) over a local network connection from either **SatDump**, **goesrecv** or **xritdecoder** and demultiplexes them into separate virtual channels, each containing a different type of image data.
The demultiplexed packets are assembled into complete files which are output as images such as the ones below.

![GK-2A Wavelengths](https://vksdr.com/bl-content/uploads/pages/ee5e126f5e958391589fea17a681d7f7/wavelengths.png)

## Getting Started
A guide for setting up the hardware and software components of a GK-2A LRIT receiver is [available on my site](https://vksdr.com/xrit-rx). It also covers the types of images that can be received, image post-processing techniques and data decryption.

<a href="https://vksdr.com/xrit-rx" target="_blank"><p align="center"><img src="https://vksdr.com/bl-content/uploads/pages/ee5e126f5e958391589fea17a681d7f7/guide-thumb-light.png" title="Receiving Images from Geostationary Weather Satellite GEO-KOMPSAT-2A"></p></a>

The [RTL-SDR Blog](https://www.rtl-sdr.com) has also [written a guide](https://www.rtl-sdr.com/rtl-sdr-com-goes-16-17-and-gk-2a-weather-satellite-reception-comprehensive-tutorial/) for setting up the hardware and software required to receive imagery from GOES-16/17 and GK-2A. Once you are able to receive the GK-2A LRIT downlink with **SatDump**, **goesrecv** or **xritdecoder**, you can begin installing and configuring **xrit-rx**.

### Installing xrit-rx
First, ensure Python version 3.7 or newer is installed by running `python3 -V` which will return something like `Python 3.8.3`.

Next, download the [latest version of **xrit-rx**](https://github.com/sam210723/xrit-rx/releases/latest) (``xrit-rx.zip``) from the Releases page, then unzip the contents to a new folder.

Some extra Python packages are required to run **xrit-rx**. Use the following command to download and install these packages:
```
pip3 install -r requirements.txt
```

[**ffmpeg**](https://www.ffmpeg.org/) is also required for processing GK-2A HRIT imagery. To install **ffmpeg** on Debian Linux (Ubuntu, Raspbian etc.) run `sudo apt install ffmpeg`. For Windows installations, [download **ffmpeg**](https://www.gyan.dev/ffmpeg/builds/ffmpeg-release-github) (`ffmpeg-essentials_build.zip`), extract `bin/ffmpeg.exe` to a new folder and [add that folder to the `PATH` environment variable](https://www.architectryan.com/2018/03/17/add-to-the-path-on-windows-10/).

Images downlinked from GK-2A are encrypted by the [Korean Meteorological Administration](https://nmsc.kma.go.kr/enhome/html/main/main.do) (KMA). Decryption keys can be downloaded from KMA's website and used with **xrit-rx**.
More information is [available in the setup guide](https://vksdr.com/xrit-rx#keys).

### Configuring xrit-rx
All user-configurable options are found in [`xrit-rx.ini`](src/xrit-rx.ini). The default configuration will work for most situations.

If **xrit-rx** is not running on the same device as **SatDump** / **goesrecv** / **xritdecoder**, the `ip` option will need to be updated with the IP address of the device running **SatDump** / **goesrecv** / **xritdecoder**.

[Click here for a full list of configuration options](#configuration-options)


## Dashboard
**xrit-rx** includes a web-based dashboard for easy monitoring and viewing of received data.
The current GK-2A LRIT schedule is also displayed on the dashboard (retrieved from [KMA NMSC](https://nmsc.kma.go.kr/enhome/html/main/main.do)).

![Dashboard](https://vksdr.com/bl-content/uploads/pages/ee5e126f5e958391589fea17a681d7f7/dashboard.png)

By default the dashboard is enabled and accessible on port <abbr title="Comes from the COMS-1/GK-2A LRIT frequency: 1692.14 MHz">1692</abbr> via HTTP (no HTTPS). These settings can be changed in the ``[dashboard]`` section of ``xrit-rx.ini``.


## Configuration Options

#### `rx` section
| Setting      | Description | Options | Default |
| ------------ | ----------- | ------- | ------- |
| `spacecraft` | Name of spacecraft being received | `GK-2A` | `GK-2A` |
| `mode`       | Type of downlink being received | `lrit` or `hrit` | `lrit` |
| `input`      | Input source | `nng`, `tcp` or `udp` | `nng` |
| `keys`       | Path to decryption key file | *Absolute or relative file path* | `EncryptionKeyMessage.bin` |

#### `output` section

| Setting   | Description | Options | Default |
| --------- | ----------- | ------- | ------- |
| `path`    | Root output path for received files | *Absolute or relative file path* | `"received"` |
| `images`  | Enable/Disable saving Image files to disk | `true` or `false` | `true` |
| `xrit`    | Enable/Disable saving xRIT files to disk | `true` or `false` | `false` |
| `enhance` | Automatically enhance LRIT IR105 images | `true` or `false` | `true` |
| `ignored` | List of virtual channels to ignore<br>Can be multiple channels (e.g. `4,5`) | `0: Full Disk`<br>`4: Alpha-numeric Text`<br>`5: Additional Data`<br> | *none* |

#### `nng` section

| Setting | Description | Options | Default |
| ------- | ----------- | ------- | ------- |
| `ip`    | IP address of a **nanomsg** data source (e.g. **SatDump** or **goesrecv**) | *Any IPv4 address* | `127.0.0.1` |
| `port`  | Server port of a **nanomsg** data source | *Any TCP port number* | `5004` |

#### `tcp` section

| Setting | Description | Options | Default |
| ------- | ----------- | ------- | ------- |
| `ip`    | IP address of a TCP data source (e.g. **xritdecoder**) | *Any IPv4 address* | `127.0.0.1` |
| `port`  | Server port of a TCP data source | *Any TCP port number* | `5001` |

#### `udp` section

| Setting | Description | Options | Default |
| ------- | ----------- | ------- | ------- |
| `ip`    | IP address to bind UDP socket to | *Any IPv4 address* | `127.0.0.1` |
| `port`  | Port number to bind UDP socket to | *Any UDP port number* | `5002` |

#### `dashboard` section

| Setting    | Description | Options | Default |
| ---------- | ----------- | ------- | ------- |
| `enabled`  | Enable/Disable dashboard server | `true` or `false` | `true` |
| `port`     | Port number for server to listen on | *Any TCP port number* | `1692` |
| `interval` | Update interval in seconds | `integer` | `1` |



## Command Line Arguments
Most of the **xrit-rx** command line arguments are only useful for development and debugging purposes. All configuration required for decoding images is done through `xrit-rx.ini` as outlined above.

```
usage: xrit-rx.py [-h] [-v] [--config CONFIG] [--file FILE] [--dump DUMP] [--no-exit]

optional arguments:
  -h, --help       Show this help message and exit
  -v, --verbose    Enable verbose console output
  --config CONFIG  Path to configuration file (*.ini)
  --file FILE      Path to VCDU packet file
  --dump DUMP      Write VCDU packets to file
  --no-exit        Pause main thread before exiting
```



## HTTP API
**xrit-rx** has a basic API accessible via HTTP primarily to support its web-based monitoring dashboard.
This may be useful for integrating **xrit-rx** with other applications.

The API only supports `GET` requests and will return either `200 OK` or `404 Not Found`.
The root endpoint located at `/api` returns information about the current xrit-rx configuration.
```json
{
  "version": 1.4,
  "spacecraft": "GK-2A",
  "downlink": "LRIT",
  "ignored": [4, 5],
  "images": true,
  "xrit": false,
  "interval": 1
}
```

The `/api/status` endpoint returns information about the current decoder state. This is used to update the dashboard interface at regular intervals.
The `image` and `xrit` values will only be present if **xrit-rx** is configured to output the respective file types (see the [Configuration Options](#configuration-options) section).

```json
{
  "vcid": 63,
  "progress": 100,
  "image": "YYYYMMDD/FD/IMG_[...].[jpg/gif/png]",
  "xrit": "YYYMMDD/FD/IMG_[...].[lrit/hrit]"
}
```

The `/api/received` endpoint enables access to received files using the paths returned by the `/api/status` endpoint. The response MIME type will change depending on the type of file requested.

For example, requesting `/api/received/20201231/FD/IMG_FD_143_IR105_20201231_2359.jpg` will return an JPEG image with MIME `image/jpeg`.

### List of Endpoints
| URL                       | Description                       | Example       | MIME                   |
| ------------------------- | --------------------------------- | ------------- | ---------------------- |
| `/api`                    | General configuration information | *see above*   | `application/json`     |
| `/api/status`             | Current demuxer status            | *see above*   | `application/json`     |
| `/api/received/[...]`     | Received data access              | *binary data* | `image/[jpeg/gif/png]` |


## Acknowledgments
  - [Lucas Teske](https://twitter.com/lucasteske) - Developer of [**Open Satellite Project**](https://github.com/opensatelliteproject) and writer of ["GOES Satellite Hunt"](https://www.teske.net.br/lucas/2016/10/goes-satellite-hunt-part-1-antenna-system/)
  - [Aang254](https://twitter.com/aang254) - Developer of [**SatDump**](https://github.com/altillimity/SatDump)
  - [Pieter Noordhuis](https://twitter.com/pnoordhuis) - Developer of [**goestools**](https://github.com/pietern/goestools)
  - [John Bell](https://twitter.com/eswnl) - Software testing and IQ recordings
  - ["kisaa"](https://github.com/kisaa) - GK-2A HRIT debugging and packet recordings
  - [@Rasiel_J](https://twitter.com/Rasiel_J) - IQ recordings


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.

<details>
<summary>Unreleased changes</summary>

**Note:** This release adds new dependencies and updates minimum versions of existing dependencies.
Run `pip3 install -r requirements.txt` to install and update these packages before using v1.4.

**Note:** This release includes changes to the configuration file format. Configuration files from previous versions will not work with v1.4.


### Added
  - Automatic LRIT IR105 image enhancement
  - Offline schedule fallback when online schedule fails
  - Dashboard schedule download failue message
  - Flag to pause main thread after processing (`--no-exit`)
  - New version notification on startup console output
  - New version notification on dashboard
  - Link to decryption key info in key error message
  - Subtle dashboard CSS animations
  - CADU to VCDU converter script

### Changed
  - Complete rewrite of `xrit-rx.py`
  - Partial rewrite of CCSDS classes (`S_PDU`)
  - Full dashboard URL displayed on startup
  - Static received data API endpoint
  - Combined status API endpoint
  - Dynamic core wait time based on downlink bitrate
  - Parse ignored VCIDs into a set
  - Remove unused legacy functions
  - Rename `goesrecv` source to `nng` (nanomsg)
  - Rename `osp` source to `tcp`
  - Use `f-strings` instead of `str.format()`

### Fixed
  - Naming convention and code quality issues
  - HTTP request handler locking up
  - Safe exit when demux core not ready
  - Relative paths when script is not in the CWD
  - Product file path manipulation using `pathlib`
  - Partial FDs not showing up on dashboard
  - Payload DES ECB block boundary alignment
  - Dashboard font fallback
  - Check config file exists before reading
  - Clipping to white in enhancement script
</details>


## [v1.3.1](https://github.com/sam210723/xrit-rx/releases/tag/v1.3.1) - Truncated Image Bugfix (2020-10-21)
Improves handling of truncated image errors for multi-segment products.

<details>
<summary>Details</summary>

### Changed
  - Clean up truncated file warning message

### Fixed
  - Handling of truncated image file exceptions in multi-segment product
</details>


## [v1.3](https://github.com/sam210723/xrit-rx/releases/tag/v1.3) - GK-2A HRIT Imagery (2020-07-12)
Added support for multi-channel HRIT imagery using [libjpeg](https://github.com/thorfdbg/libjpeg) for JPEG2000 conversion.

<details>
<summary>Details</summary>

### Added
  - Support for HRIT full disk imagery (see [HRIT Decoding - J2K and missing CPPDUs](https://github.com/sam210723/xrit-rx/issues/15))
  - HRIT image processing tool ([tools\hrit-img.py](https://github.com/sam210723/xrit-rx/blob/master/src/tools/hrit-img.py))
  - JP2 (JPEG2000) to PPM converstion using [libjpeg](https://github.com/thorfdbg/libjpeg) (licensed under [GPLv3](https://github.com/thorfdbg/libjpeg/blob/master/README.license.gpl))
  - UDP socket for receiving VCDUs from hardware modems such as the ETRA D8L
  - to_hex() debugging utility method
  - Link to GitHub release page on dashboard

### Changed
  - Indicate multi-segment progress per-wavelength rather than for the entire product
  - Using `pathlib` over `os` module for some file operations
  - Include spacecraft and downlink in dashboard schedule title
  - Renamed "Last Image" dashboard block to "Latest Image"
  - Renamed `last/image` and `last/xrit` API endpoints to `latest/image` and `latest/xrit`

### Fixed
  - TP_File triggering with M_PDU header offset ([relevant issue comment](https://github.com/sam210723/xrit-rx/issues/15#issuecomment-643079493))
  - Output directory checking
  - Handling of safe exit cases
  - Handling of `PIL.UnidentifiedImageError`
  - Handling of missing configuration sections and options
</details>


## [v1.2](https://github.com/sam210723/xrit-rx/releases/tag/v1.2) - Monitoring Dashboard (2020-05-14)
Added a monitoring dashboard and JSON API running from a built in web server.

<details>
<summary>Details</summary>

### Added
  - Web-based monitoring dashboard
  - JSON API for updating dashboard
  - Access of received data over HTTP
</details>


## [v1.1](https://github.com/sam210723/xrit-rx/releases/tag/v1.1) - Automatic Product Generation (2020-02-14)
Products (images/text) are now output directly from xrit-rx rather than relying on ``lrit-img.py`` and ``lrit-add.py``.
Progress bar for multi-segment images and colour-coded success/failure messages were also added.

<details>
<summary>Details</summary>

### Added
  - Output products (images/text) directly from demuxer
  - Transparent enhanced image output option
  - Added check for encrypted LRIT files in ``lrit-img.py`` and ``lrit-add.py``
  - Output file type options (Image or xRIT files)
  - Demuxer configuration tuple
  - Channel handler configuration tuple
  - Detect GK-2A LRIT Daily Operation Plan
  - Console output colours
  - Progress bar for multi-segment images

### Changed
  - Default key file name **(check when upgrading from an old version)**
  - ``keymsg-decrypt.py`` output file name
  - Disable product output if no keys loaded
  - Write single fill VCDU to packet file on VCID change
  - Rename FILL packets to IDLE packets

### Fixed
  - Missing TrueType font exception
</details>


## [v1.0.3](https://github.com/sam210723/xrit-rx/releases/tag/v1.0.3) - Infrared Enhancement Tool (2019-12-09)
Added work-around for ``COMSFOG`` and ``COMSIR1`` transmission issue, an infrared imagery enhancement tool, and fixed some demuxer bugs.

<details>
<summary>Details</summary>

### Added
  - IR enhancement tool ([tools\enhance-ir.py](https://github.com/sam210723/xrit-rx/blob/master/src/tools/enhance-ir.py))
  - Extra demuxer info in verbose mode

### Changed
  - Write incomplete TP_Files to disk on VCID change ([COMSFOG / COMSIR1 issue](https://github.com/sam210723/xrit-rx/issues/5))
  - Clear xRIT key header after file is decrypted (avoids double-decryption)

### Fixed
  - Free-running loop while demuxing a file

src/xrit-rx.py
"""
xrit-rx.py
https://github.com/sam210723/xrit-rx

Receive images from geostationary weather satellites
"""

import argparse
import colorama
from   collections import namedtuple
import configparser
import os
from   pathlib import Path
import requests
import socket
import struct
import sys
import time

from   demuxer import Demuxer
from   dash import Dashboard


class Main:
    def __init__(self):
        print("┌──────────────────────────────────────────────┐\n"
              "│                   xrit-rx                    │\n"
              "│         LRIT/HRIT Downlink Processor         │\n"
              "├──────────────────────────────────────────────┤\n"
              "│     @sam210723         vksdr.com/xrit-rx     │\n"
              "└──────────────────────────────────────────────┘\n")
        
        # Initialise Colorama
        colorama.init(autoreset=True)

        # Check minimum Python version requirement
        if sys.version_info[1] < 7:
            self.log("Python v3.7 or newer is required to run xrit-rx", style="error")
            exit(1)

        # Set instance variables
        self.demuxer = None         # Demuxer class instance
        self.dashboard = None       # Dashboard class instance
        self.keys = None            # Encryption key list
        self.packet_file = None     # Packet input file
        self.dump_file = None       # Packet output file
        self.version = "1.4"        # Application version

        # Information dictionary
        self.info = {
            "GK-2A": {
                "name": "GEO-KOMPSAT-2A (GK-2A)",
                "LRIT": [1692.14, "64 kbps", 64e3],
                "HRIT": [1695.4,   "3 Mbps",  3e6],
                "SCID": 195,
                "VCID": {
                     0: "FULL DISK",
                     4: "ALPHA-NUMERIC TEXT",
                     5: "ADDITIONAL DATA",
                    63: "IDLE"
                }
            }
        }

        # Configure xrit-rx
        self.configure()

        # Setup input source
        self.setup_input()


    def configure(self):
        """
        Configures xrit-rx (args, input file, output file, cwd, config file, keys, updates)
        """

        # Get command line arguments
        argp = argparse.ArgumentParser()
        argp.add_argument("-v", "--verbose", action="store_true", help="Enable verbose console output (only useful for debugging)")
        argp.add_argument("--config",        action="store",      help="Path to configuration file (*.ini)", default="xrit-rx.ini")
        argp.add_argument("--file",          action="store",      help="Path to VCDU packet file")
        argp.add_argument("--dump",          action="store",      help="Write VCDU packets to file (only useful for debugging)")
        argp.add_argument("--no-exit",       action="store_true", help="Pause main thread before exiting (only useful for debugging)")
        self.args = argp.parse_args()

        # Open input packet file
        if self.args.file:
            self.args.file = Path(self.args.file)

            # Check packet file exists
            if not self.args.file.is_file():
                self.log(f"PACKET FILE \"{self.args.file.absolute()}\" DOES NOT EXIST", style="error")
                self.stop(code=1)

            self.packet_file = open(self.args.file, "rb")

        # Open packet dump file
        if self.args.dump:
            self.args.dump = Path(self.args.dump)
            
            try:
                self.dump_file = open(self.args.dump, "wb+")
            except OSError as e:
                self.log(f"UNABLE TO OPEN PACKET DUMP FILE\"{self.args.dump.absolute()}\"", style="error")
                self.stop(code=1)

        # Change working directory to script location
        os.chdir(Path(__file__).parent.absolute())

        # Load configuration file
        config_path = Path(self.args.config)
        if config_path.is_file():
            confp = configparser.ConfigParser()
            confp.read(config_path)
            self.config = confp._sections
        else:
            self.log(f"CONFIGURATION FILE \"{self.args.config}\" DOES NOT EXIST", style="error")
            self.stop(code=1)
        
        # Create path objects
        self.config['rx']['keys'] = Path(self.config['rx']['keys'])
        self.config['rx']['mode'] = self.config['rx']['mode'].upper()
        self.config['output']['path'] = Path(self.config['output']['path']) / self.config['rx']['mode']
        self.config['output']['path'].mkdir(parents=True, exist_ok=True)

        # Parse boolean options
        self.config['output']['images'] = self.config['output']['images'] == "true"
        self.config['output']['xrit'] = self.config['output']['xrit'] == "true"
        self.config['output']['enhance'] = self.config['output']['enhance'] == "true"
        self.config['dashboard']['enabled'] = self.config['dashboard']['enabled'] == "true"

        # Parse ignored channel list
        if self.config['output']['ignored']:
            self.config['output']['ignored'] = self.config['output']['ignored'].split(',')
            self.config['output']['ignored'] = [int(c) for c in self.config['output']['ignored']]
            self.config['output']['ignored'] = set(self.config['output']['ignored'])
        else:
            self.config['output']['ignored'] = set()

        # Create ignored VCID string
        spacecraft = self.config['rx']['spacecraft']
        ignored = ", ".join(f"{c} ({self.info[spacecraft]['VCID'][c]})" for c in self.config['output']['ignored'])

        # Limit dashboard refresh rate
        self.config['dashboard']['interval'] = round(float(self.config['dashboard']['interval']), 1)
        self.config['dashboard']['interval'] = max(1, self.config['dashboard']['interval'])

        # Check spacecraft is valid
        if spacecraft not in self.info:
            self.log(f"INVALID SPACECRAFT \"{spacecraft}\"", style="error")
            self.stop(code=1)
        
        # Check downlink is valid
        downlink = self.config['rx']['mode']
        if downlink not in self.info[spacecraft]:
            self.log(f"INVALID DOWNLINK \"{downlink}\"", style="error")
            self.stop(code=1)
        
        # Check input type is valid
        if self.config['rx']['input'] not in ['nng', 'tcp', 'udp']:
            self.log(f"INVALID INPUT TYPE \"{self.config['rx']['input']}\"", style="error")
            self.stop(code=1)
        
        # Get dashboard URL
        try:
            s = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
            s.settimeout(1)
            s.connect(("10.255.255.255", 1))
            ip = s.getsockname()[0]
            s.close()
        except socket.error:
            ip = "127.0.0.1"
        dashboard_url = f"http://{ip}:{self.config['dashboard']['port']}"

        # Get input path
        if not self.args.file:
            input_path =  f"{self.config['rx']['input']}"
            input_path += f" ({self.config[self.config['rx']['input']]['ip']}:{self.config[self.config['rx']['input']]['port']})"
        else:
            input_path = self.args.file.name
            self.config['rx']['input'] = "file"
        
        # Print configuration info
        self.log(f"SPACECRAFT:   {self.info[spacecraft]['name'] if spacecraft in self.info else spacecraft}")
        self.log(f"DOWNLINK:     {downlink.upper()} ({self.info[spacecraft][downlink][0]} MHz, {self.info[spacecraft][downlink][1]})")
        self.log(f"INPUT:        {input_path}")
        self.log(f"OUTPUT:       {self.config['output']['path'].absolute()}")
        self.log(f"KEY FILE:     {self.config['rx']['keys'].name}")
        self.log(f"IGNORED:      {ignored if len(ignored) > 0 else 'None'}")
        self.log(f"DASHBOARD:    {dashboard_url if self.config['dashboard']['enabled'] else 'DISABLED'}")
        self.log(f"VERSION:      v{self.version}\n")

        # Load encryption keys
        self.keys = self.load_keys(self.config['rx']['keys'])

        # Dump file status
        if self.args.dump: self.log(f"WRITING PACKETS TO \"{self.args.dump.absolute()}\"", style="ok")

        # Check for new version on GitHub
        try:
            r = requests.get("https://api.github.com/repos/sam210723/xrit-rx/releases/latest")
            if r.status_code == 200:
                latest_tag = r.json()['tag_name']
                if f"v{self.version}" != latest_tag:
                    self.log(f"\nA new version of xrit-rx is available on GitHub", style="ok")
                    self.log("https://github.com/sam210723/xrit-rx/releases/latest\n", style="ok")
        except Exception: pass


    def load_keys(self, path):
        """
        Loads encryption keys from key file
        """

        # Check key file exists
        if not path.is_file():
            # Only output encrypted xRIT files
            self.config['output']['images'] = False
            self.config['output']['xrit'] = True

            self.log(f"KEY FILE \"{path.absolute()}\" DOES NOT EXIST", style="error")
            self.log(f"Encrypted xRIT files will be saved to disk", style="error")
            self.log(f"See https://vksdr.com/xrit-rx#keys for more information\n", style="error")
            return {}
        
        # Load key file
        key_file = open(path, "rb")

        # Get number of keys
        key_count = int.from_bytes(key_file.read(2), byteorder="big")

        # Loop through keys
        keys = {}
        for _ in range(key_count):
            # Get key index and key value
            key = struct.unpack(">H8s", key_file.read(10))

            # Add key to dict
            keys[key[0]] = key[1]

        key_file.close()
        self.log("DECRYPTION KEYS LOADED", style="ok")
        return keys


    def setup_input(self):
        """
        Sets up the selected input source
        """

        if self.config['rx']['input'] == "nng":
            # Create socket and address
            self.socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
            addr = (self.config['nng']['ip'], int(self.config['nng']['port']))

            # Connect socket
            print(f"Connecting to {addr[0]}:{addr[1]}...", end='', flush=True)
            self.connect_socket(self.socket, addr)

            # Setup nanomsg publisher in goesrecv
            self.socket.send(b'\x00\x53\x50\x00\x00\x21\x00\x00')
            if self.socket.recv(8) != b'\x00\x53\x50\x00\x00\x20\x00\x00':
                self.log("ERROR CONFIGURING NANOMSG", style="error")
                self.stop(code=1)

        elif self.config['rx']['input'] == "tcp":
            # Create socket and address
            self.socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
            addr = (self.config['tcp']['ip'], int(self.config['tcp']['port']))

            # Connect socket
            print(f"Connecting to {addr[0]}:{addr[1]}...", end='', flush=True)
            self.connect_socket(self.socket, addr)

        elif self.config['rx']['input'] == "udp":
            # Create socket and address
            self.socket = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
            addr = (self.config['udp']['ip'], int(self.config['udp']['port']))

            # Bind socket
            print(f"Binding UDP socket to {addr[0]}:{addr[1]}...", end='', flush=True)
            try:
                self.socket.bind(addr)
                self.log("SUCCESS", style="ok")
            except socket.error as e:
                self.log("FAILED", style="error")
                self.log(e)
                self.stop(code=1)


    def connect_socket(self, s, addr):
        """
        Connect a socket and handle exceptions
        """

        try:
            s.connect(addr)
            self.log("CONNECTED", style="ok")
        except socket.error as e:
            if e.errno == 10061: self.log("CONNECTION REFUSED", style="error")
            self.log(e)
            self.stop(code=1)


    def start(self):
        """
        Starts xrit-rx components
        """

        # Create demuxer instance
        demux_config = namedtuple('demux_config', 'spacecraft downlink verbose dump output images xrit enhance ignored keys info')
        self.demuxer = Demuxer(
            demux_config(
                self.config['rx']['spacecraft'],
                self.config['rx']['mode'],
                self.args.verbose,
                self.dump_file,
                self.config['output']['path'],
                self.config['output']['images'],
                self.config['output']['xrit'],
                self.config['output']['enhance'],
                self.config['output']['ignored'],
                self.keys,
                self.info
            )
        )

        # Create dashboard instance
        if self.config['dashboard']['enabled']:
            dash_config = namedtuple('dash_config', 'port interval spacecraft downlink output images xrit ignored version')
            self.dashboard = Dashboard(
                dash_config(
                    self.config['dashboard']['port'],
                    self.config['dashboard']['interval'],
                    self.config['rx']['spacecraft'],
                    self.config['rx']['mode'],
                    self.config['output']['path'],
                    self.config['output']['images'],
                    self.config['output']['xrit'],
                    self.config['output']['ignored'],
                    self.version
                ),
                self.demuxer
            )

        # Check demuxer thread is ready
        if not self.demuxer.core_ready:
            self.log("DEMUXER CORE THREAD FAILED TO START", style="error")
            self.stop(code=1)

        print("───────────────────────────────────────────────────────────────\n")

        # Get processing start time
        self.start_time = time.time()

        # Enter main loop
        self.loop()


    def loop(self):
        """
        Handles data from the selected input source
        """

        # Packet length (1 VCDU)
        buflen = 892

        while True:
            if self.config['rx']['input'] == "nng":
                # Get packet from nanomsg source
                try:
                    data = self.socket.recv(buflen + 8)
                except ConnectionResetError:
                    self.log("LOST CONNECTION TO NANOMSG SOURCE", style="error")
                    self.stop(code=1)

                # Push packet to demuxer
                if len(data) == buflen + 8: self.demuxer.push(data[8:])
            
            elif self.config['rx']['input'] == "tcp":
                # Get packet from TCP source
                try:
                    data = self.socket.recv(buflen)
                except ConnectionResetError:
                    self.log("LOST CONNECTION TO TCP SOURCE", style="error")
                    self.stop(code=1)

                # Push packet to demuxer
                if len(data) == buflen: self.demuxer.push(data)

            elif self.config['rx']['input'] == "udp":
                # Get packet from UDP source
                try:
                    data, _ = self.socket.recvfrom(buflen)
                except Exception as e:
                    self.log(e, style="error")
                    self.stop(code=1)
                
                # Push packet to demuxer
                if len(data) == buflen: self.demuxer.push(data)

            elif self.config['rx']['input'] == "file":
                if not self.packet_file.closed:
                    # Read packet from file
                    data = self.packet_file.read(buflen)

                    # No more data to read from file
                    if data == b'':
                        self.packet_file.close()
                        
                        # Append single fill VCDU (VCID 63)
                        # Triggers TP_File processing inside channel handlers
                        # by changing the currently active VCID
                        self.demuxer.push(b'\x70\xFF\x00\x00\x00\x00')
                        continue
                    
                    # Push packet to demuxer
                    self.demuxer.push(data)
                else:
                    # Demuxer has all packets from file in its queue
                    # Wait for processing to finish
                    if self.demuxer.complete():
                        run_time = round(time.time() - self.start_time, 3)
                        self.log(f"\nPROCESSED FILE IN {run_time:.03f}s", style="ok")
                        self.stop()
                    else:
                        # Limit loop speed when waiting for demuxer to finish processing
                        time.sleep(0.5)


    def stop(self, msg=True, code=0):
        """
        Stops xrit-rx gracefully
        """

        # Close dump file
        if self.dump_file: self.dump_file.close()

        # Keep running until keyboard interrupt
        if self.args.no_exit:
            self.log("PAUSING MAIN THREAD (--no-exit)", style="ok")
            try:
                while True: time.sleep(0.5)
            except KeyboardInterrupt: pass
        
        # Stop child threads
        if self.demuxer: self.demuxer.stop()
        if self.dashboard: self.dashboard.stop()

        # Show message and exit
        if msg: self.log("Exiting...")
        exit(code)


    def log(self, msg, style="none"):
        """
        Writes to console
        """

        # Colorama styles
        styles = {
            "none":   "",
            "ok":    f"{colorama.Fore.GREEN}{colorama.Style.BRIGHT}",
            "error": f"{colorama.Fore.WHITE}{colorama.Back.RED}{colorama.Style.BRIGHT}"
        }
        
        print(f"{styles[style]}{msg}")


# Initialise xrit-rx
instance = Main()
try:
    # Start xrit-rx
    instance.start()
except KeyboardInterrupt:
    # Exit on keyboard interrupt
    instance.stop()
